core/src/Kokkos_Parallel_Reduce.hpp

```diff
@@ -30,37 +30,54 @@ static_assert(false,
 
 namespace Kokkos {
 
+/** \brief Class offering functionalities common to all reducers
+ *
+ * In order to be a valid reducer, a class must implement the functions and
+ * define the types offered in this class.
+ * To facilitate implementation, a new reducer class can simply inherit from
+ * BaseReducer.
+ */
+namespace Impl {
 template <class Scalar, class Space>
-struct Sum {
+struct BaseReducer {
  public:
-  // Required
-  using reducer    = Sum<Scalar, Space>;
-  using value_type = std::remove_cv_t<Scalar>;
-  static_assert(!std::is_pointer_v<value_type> && !std::is_array_v<value_type>);
-
+  // Following types need to be available for the reducer to be valid
+  using value_type       = std::remove_cv_t<Scalar>;
   using result_view_type = Kokkos::View<value_type, Space>;
 
- private:
+  static_assert(!std::is_pointer_v<value_type> && !std::is_array_v<value_type>);
+
+ protected:
+  // Contains the value of the reduction
   result_view_type value;
+  // Whether the reducer returns its value through a Kokkos::View or a scalar
   bool references_scalar_v;
 
  public:
+  // Construct from a scalar value
   KOKKOS_INLINE_FUNCTION
-  Sum(value_type& value_) : value(&value_), references_scalar_v(true) {}
+  BaseReducer(value_type& value_) : value(&value_), references_scalar_v(true) {}
 
+  // Construct from a View
   KOKKOS_INLINE_FUNCTION
-  Sum(const result_view_type& value_)
+  BaseReducer(const result_view_type& value_)
       : value(value_), references_scalar_v(false) {}
 
-  // Required
-  KOKKOS_INLINE_FUNCTION
-  void join(value_type& dest, const value_type& src) const { dest += src; }
+  /* Reducers also need to implement the two following functions:
+   *
+   * KOKKOS_INLINE_FUNCTION
+   * void join(value_type& dest, const value_type& src) const {
+   *    // Do reduction here
+   * }
 
-  KOKKOS_INLINE_FUNCTION
-  void init(value_type& val) const {
-    val = reduction_identity<value_type>::sum();
-  }
+   * KOKKOS_INLINE_FUNCTION
+   * void init(value_type& val) const {
+   *   // Set the initial value for the reduction variable here
+   *   // (for instance FLOAT_MIN if searching for the max).
+   * }
+   */
```

Comment:
Maybe we should use a CRTP-type approach and implement these functions here and call `init_impl` and `join_impl` defined in the derived classes.
Note that implementing `init` is optional but we could require it for classes derived fromt his one.